text/0000-relax-const-restrictions.md

```diff
@@ -0,0 +1,82 @@
+
+Rust code often contains debug assertions or preconditions that must be upheld but are too expensive to check in release mode. It is desirable to also check these preconditions during constant evaluation (for example with a `debug_or_const_assert!` macro). This is unsound under the old rules, as this would be different behavior during const evaluation in release mode. This RFC allows such debug assertions to also run during constant evaluation (but does not propose this itself).
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+This RFC allows `const fn` to exhibit different behavior during constant evaluation and runtime. While such a difference is often undesirable, it is not considered to be undefined behavior.
+
+If a `const fn` is able to detect whether it has been called during constant evaluation or at runtime (either through an intrinsic or a future language feature), then it is allowed to exhibit different behavior. Also, a `const fn` called at runtime can do anything a normal function can do, with no additional restrictions applied to it. It could open a file, call a system randomness API or gracefully exit the program. 
+
+At the time of writing, there is no way for a function to detect whether it was called at runtime or during constant evaluation in stable Rust and this RFC is not concerned with adding any, but it encourages future RFCs for adding this capability.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+This RFC introduces a new concept called the "constness-context" of an execution of a function. A `const fn` is executed in a const constness-context if it was called in a context where only `const fn` are allowed (like `static` initializers) or inside another `const fn` that was executed in a const constness-context. All other calls to a `const fn` (for example in `main`) are in a runtime constness-context.
+
+We can therefore say that code can either be:
+- Always in a const constness-context (for example `static` initializers, array lengths...)
+- Maybe in a const constness-context (`const fn`), where the constness-context can differ between calls depending on the call-site
+- Always in a runtime const constness-context (non-`const` functions)
+
+A `const fn` is now allowed to exhibit different behavior depending on it being called in a const or runtime constness-context. Language features and standard library functions may also differ in behavior depending on the constness-context they have been used or called in.
+
+This makes the constness-context of a function observable behavior.
+
+If a `const fn` is called in a runtime constness-context, no additional restrictions are applied to it, and it may do anything a non-`const fn` can (for example, calling into FFI).
+
+A `const fn` being called in a const constness-context will still be required to be deterministic, as this is required for type systme soundness.
```

Comment:
Maybe clarify that this is enforced by the compiler, so even unsafe code has no way of breaking this requirement.